packages/juce-blueprint/src/components/Button.tsx

```diff
@@ -0,0 +1,73 @@
+import React, { useState } from 'react';
+
+import { View } from './View'
+import { SyntheticMouseEvent } from '../lib/SyntheticEvents'
+
+export interface ButtonProps {
+    onClick: () => void;
+    onMouseDown?: (e: SyntheticMouseEvent) => void;
+    onMouseUp?: (e: SyntheticMouseEvent) => void;
+}
+
+/**
+ * A simple Button component which can be used as a building block
+ * for more complex button types.
+ *
+ * @example
+ *
+ * <Button onClick={() => { console.log("clicked"); }} {...styles.button}>
+ *   <Text {...styles.text}>
+       Hello World
+ *   </Text>
+ * </Button>
+ *
+ * const styles = {
+ *   text: {
+ *     'font-size': 18.0,
+ *     'line-spacing': 1.6,
+ *     'color': 'ff626262'
+ *   },
+ *   button: {
+ *     'justify-content': 'center',
+ *     'align-items': 'center',
+ *     'width': '100%',
+ *     'height': '100%',
+ *     'border-radius': 5.0,
+ *     'border-width': 2.0,
+ *     'border-color': 'ff626262'
+ *   }
+ * };
+ *
+ */
+export const Button = (props: React.PropsWithChildren<ButtonProps>) => {
+    const { onClick, onMouseDown, onMouseUp, ...other } = props;
+
+    const [down, setDown] = useState(false);
+
+    const handleDown = (e: SyntheticMouseEvent) => {
+        if (typeof onMouseDown === 'function')
+            onMouseDown.call(null, e);
+
+        setDown(true);
+    }
+
+    const handleUp = (e: SyntheticMouseEvent) => {
+        if (typeof onMouseUp === 'function')
+            onMouseUp.call(null, e);
+
+        setDown(false);
+        onClick();
```

Comment:
Interesting here, I'm just thinking of this now. I don't think `Button` should be the one sending the `onClick` up the tree. I think we could just remove the handling of `onClick` altogether here.

That said, the `ViewInstance` inside the js renderer should send click events with a SyntheticMouseEvent as in here: https://developer.mozilla.org/en-US/docs/Web/API/Element/click_event (see https://github.com/nick-thompson/blueprint/blob/master/packages/juce-blueprint/src/lib/BlueprintBackend.ts#L99)

For now, want to just remove the handling of `onClick` here and queue the ViewInstance work for later? Or would it be helpful to tackle now?

Reply:
@nick-thompson, I think it makes sense to add this now and I see your point. 

So essentially we just want to switch on `onMouseUp` as the event in the `if (SyntheticEvents.isMouseEventHandler(propKey))` check and then check if `onClick` exists as a prop and call ? 

i.e. sit closer to React's proper SynteticEvent behaviour?

Reply:
Oh wait it doesn't even need to do that does it. C++ should send the click event up through that chain. Fool.

Reply:
Out of interest. Should our view event names match React's ?

`onClick onContextMenu onDoubleClick onDrag onDragEnd onDragEnter onDragExit
onDragLeave onDragOver onDragStart onDrop onMouseDown onMouseEnter onMouseLeave
onMouseMove onMouseOut onMouseOver onMouseUp`

i.e. we use `onMouseDoubleClick` should we be using `onDoubleClick` ?

Reply:
Yea, let's definitely aim for parity with React's event names. And now is the time for breaking changes :) Glad we're doing this now, it'll be nice to minimize expected breaking changes once we announce beta.

> So essentially we just want to switch on onMouseUp as the event in the if (SyntheticEvents.isMouseEventHandler(propKey)) check and then check if onClick exists as a prop and call ?
>
> i.e. sit closer to React's proper SynteticEvent behaviour?

Exactly. With `onClick` though there's slightly more complicated behavior. I'll elaborate a bit here on how this part works to hopefully clarify why this one's complicated and what we can do about it. Feel free to skip over this paragraph if you already know it :). So, blueprint ships only a few core components (View, Image, Canvas, ScrollView... etc), and all of these components natively derive from `blueprint::View`. So, no matter what your application looks like, by the time it hits the BlueprintRenderer, pretty much everything is just a `View`. To track the rendered instances, we make this `ViewInstance` class internally. Here, when we see that you're trying to supply an `onMouseDown`, handler, we hijack it. We put our _own_ `onMouseDown` handler over the native boundary, and hang onto the one you gave us. Then, from the native side, we listen for juce's Component mouse callbacks and try to invoke the `onMouseDown` that we handed over to native: https://github.com/nick-thompson/blueprint/blob/master/blueprint/core/blueprint_View.cpp#L196. When that happens, we catch the call, run the `onMouseDown` that you gave us, but then apply our own behavior for bubbling this up the `ViewInstance` tree, as you see in React/React-Native and normal web DOM stuff. So in the naive case, if juce had an `onClick` method of `juce::Component`, we could just run the same play to wire this up for our `onClick` prop.

Now the reason this one in particular is complicated, is because an `onClick` is (according to the spec) only fired when we've seen both a mouse down _and_ a mouse up. Imagine your app running on your desktop, clicking somewhere on your desktop (not on your app), holding the button down, moving your mouse over your app, and releasing the mouse button. We'll fire a mouseUp, but not a mouseDown (we didn't see the mouseDown, your desktop did). So we should not fire an onClick. More complicated: per the spec, we should fire an onClick from a parent `ViewInstance` even when one of the children saw the mouse down.

So one way we could implement this is to give _every_ `ViewInstance` a default `onMouseDown`/`onMouseUp` and let them each do their own bookkeeping to see if they should fire an `onClick`. That could get costly, and we really don't want all those functions getting set onto the native side. Another way would be to create a top level "book keeper," maintain our existing functionality, but have our `mouseDown`/`mouseUp` handlers tell the book keeper "hey I just got a mouse down/up." Then the book keeper identifies when to run an `onClick`, and invokes its own bubble mechanism up the appropriate `ViewInstance` subtree. This would be the preferred way, imo, because it creates the least overhead. (FWIW this is, as far as I've discerned from a very quick look over react's internal implementation, how React's synthetic event library works too: with a top level listener and this notion of "replaying" events over the correct child nodes– https://github.com/facebook/react/blob/master/packages/react-dom/src/events/ReactDOMEventReplaying.js)

What do you think?

Reply:
@nick-thompson , So I'd like to try tackling this soon. 

I just want to clarify my understanding on this. 

Basically we're saying that in `ViewInstance` if a `setProperty` call comes through for a mouse event then we can check if that event is `onMouseDown` / `onMouseUp` and then call a function to register that event for the given `ViewInstance` with some `EventBookKeeper` class ?

I.e. 

```
// in setProperty
if (SyntheticEvents.isMouseEventHandler([propKey)) {
  if (isMouseUpOrDown(propKey)) {
    eventBookKeeper.notify(this._id, propKey);
  }
}
```

In the above `eventBookKeeper` (or whatever we call it) would store the previous id/event pair and on notify if the previous event belongs to the same view instance fetch the instance from the root container and call `instance.bubbleViewEvent` (or something along those lines). 

I'm sure the event tracking will get more complex than this as we go along and we'll need to define how something like a `mouseDrag` within a single view followed by a release should behave etc (imagining some sort of id -> event stack map or something). 

Is this crude description at all in line with what you're describing?

Reply:
Yea, so this is indeed fairly tricky.

In general you have the right idea: if we equip every single `ViewInstance` with a default handler (currently we're using that noop function https://github.com/nick-thompson/blueprint/blob/master/packages/juce-blueprint/src/lib/BlueprintBackend.ts#L50), and use that wrapped callback to notify our bookkeeper, then the bookkeeper will be able to definitively say "we saw a mouse down over here, and a mouseup over here, which means we should run a click event from here up." In that case, yea, I think the code snippet you posted above is about right.

Off the bat, I wonder if this is an expensive way to do it. For example, in a large application, do we really want default noop functions for several mouse events on _every_ view instance just so that we can bubble? (Remember, all of those prop functions go over the native boundary). Or would it perform better to have a more generic "dispatchViewEvent" that goes directly to the bookkeeper, who is then responsible for replaying the event up the tree from the child emitting the event. That would simplify the bookkeeping and bubbling too, since we don't need to wrap every prop function with some wrapper code for bubbling. This is actually how the view event stuff was first written, and I scooped it out when prop functions became supported. Now that we're considering proper event bubbling, I wonder if that was a wrong move :)

Anyways! I think we should start with the proposed method two paragraphs up, with the goal of just correctly satisfying the onclick event. If we have that behavior correct, then user code would never have to change if we reimplemented the view stuff along the lines of the proposal one paragrah up.

Reply:
@nick-thompson , Yeah sorry just whacked that example in without thinking but of course it wouldn't happen in that exact location as you'd have to persist some callback/handler. Gotcha. 

Having `dispatchViewEvent` push events into some "book keeping" class does sound like a sensible way to do this. I agree on the overhead concerns and I'm all for anything that helps to reduce memory usage inside duktape etc. 

Agree'd though. I can implement the first approach just to satisfy `onClick` and then allow us to get `Button` and the like into the core components. 

I'll take a quick look over the `dispatchViewEvent` approach just in case it looks like something I can get out of the way in a reasonable amount of time!